File: ontrack-extension-git/src/main/java/net/nemerosa/ontrack/extension/git/service/GitService.java
package net.nemerosa.ontrack.extension.git.service;

import net.nemerosa.ontrack.extension.api.model.BuildDiffRequest;
import net.nemerosa.ontrack.extension.git.model.*;
import net.nemerosa.ontrack.extension.git.property.GitBranchConfigurationProperty;
import net.nemerosa.ontrack.extension.scm.service.SCMService;
import net.nemerosa.ontrack.model.Ack;
import net.nemerosa.ontrack.model.structure.Branch;
import net.nemerosa.ontrack.model.structure.ID;
import net.nemerosa.ontrack.model.structure.Project;
import org.eclipse.jgit.revwalk.RevCommit;

import java.util.List;
import java.util.Optional;
import java.util.concurrent.Future;
import java.util.function.BiConsumer;
import java.util.function.Predicate;

public interface GitService extends SCMService {

    /**
     * Tests if a branch is correctly configured for Git.
     */
    boolean isBranchConfiguredForGit(Branch branch);

    /**
     * Gets the configuration for a project.
     */
    Optional<GitConfiguration> getProjectConfiguration(Project project);

    /**
     * Gets the configuration for a branch
     *
     * @param branch Branch to check
     * @return Configuration.
     */
    Optional<GitBranchConfiguration> getBranchConfiguration(Branch branch);

    /**
     * Launches the build/tag synchronisation for a branch
     */
    Optional<Future<?>> launchBuildSync(ID branchId, boolean synchronous);

    /**
     * Change log
     */
    GitChangeLog changeLog(BuildDiffRequest request);

    /**
     * Change log commits
     */
    GitChangeLogCommits getChangeLogCommits(GitChangeLog changeLog);

    /**
     * Change log issues
     */
    GitChangeLogIssues getChangeLogIssues(GitChangeLog changeLog);

    /**
     * Change log files
     */
    GitChangeLogFiles getChangeLogFiles(GitChangeLog changeLog);

    /**
     * Loops over each correctly configured project.
     */
    void forEachConfiguredProject(BiConsumer<Project, GitConfiguration> consumer);

    /**
     * Loops over each correctly configured branch. Branch template definitions are excluded.
     */
    void forEachConfiguredBranch(BiConsumer<Branch, GitBranchConfiguration> consumer);

    /**
     * Gets information about an issue in a Git-configured branch
     */
    OntrackGitIssueInfo getIssueInfo(ID branchId, String key);

    /**
     * Looks up a commit in the given <code>configuration</code>.
     *
     * @param id Commit long or short ID
     * @return The content of a commit if it exists, empty otherwise.
     */
    Optional<GitUICommit> lookupCommit(GitConfiguration configuration, String id);

    /**
     * Gets information about a commit in a Git-configured project.
     */
    OntrackGitCommitInfo getCommitProjectInfo(ID projectId, String commit);

    /**
     * Gets the list of remote branches, as defined under <code>ref/heads</code>.
     */
    List<String> getRemoteBranches(GitConfiguration gitConfiguration);

    /**
     * Gets a diff on a list of file changes, filtering the changes using ANT-like patterns
     */
    String diff(GitChangeLog changeLog, List<String> patterns);

    /**
     * Synchronises the Git repository attached to this project.
     *
     * @param project Project
     * @param request Sync request
     * @return Result. The synchronisation will occur asynchronously, but the acknowledgment returns
     * if the project did contain a Git configuration or not.
     */
    Ack projectSync(Project project, GitSynchronisationRequest request);

    /**
     * Synchronises the Git repository attached to this configuration.
     *
     * @param gitConfiguration Configuration to sync
     * @param request          Sync request
     * @return Result. The synchronisation will occur asynchronously, but the acknowledgment returns if the
     * synchronisation was actually launched.
     */
    Optional<Future<?>> sync(GitConfiguration gitConfiguration, GitSynchronisationRequest request);

    /**
     * Gets the Git synchronisation information.
     *
     * @param project Project configured for Git
     * @return Synchronisation information
     */
    GitSynchronisationInfo getProjectGitSyncInfo(Project project);

    void scheduleGitBuildSync(Branch branch, GitBranchConfigurationProperty property);

    void unscheduleGitBuildSync(Branch branch, GitBranchConfigurationProperty property);

    /**
     * Checks the log history and returns <code>true</code> if the token can be found.
     *
     * @param branchConfiguration Git configuration
     * @param token               Expression to be searched for
     * @return Result of the search
     */
    boolean isPatternFound(GitBranchConfiguration branchConfiguration, String token);
}

File: ontrack-extension-git/src/main/java/net/nemerosa/ontrack/extension/git/service/GitService.kt
package net.nemerosa.ontrack.extension.git.service

import net.nemerosa.ontrack.extension.git.model.*
import net.nemerosa.ontrack.extension.git.property.GitBranchConfigurationProperty
import net.nemerosa.ontrack.extension.issues.export.ExportFormat
import net.nemerosa.ontrack.extension.scm.service.SCMService
import net.nemerosa.ontrack.git.GitRepositoryClient
import net.nemerosa.ontrack.git.model.GitCommit
import net.nemerosa.ontrack.job.JobCategory
import net.nemerosa.ontrack.job.JobRunListener
import net.nemerosa.ontrack.model.Ack
import net.nemerosa.ontrack.model.structure.Branch
import net.nemerosa.ontrack.model.structure.Build
import net.nemerosa.ontrack.model.structure.ID
import net.nemerosa.ontrack.model.structure.Project
import java.util.concurrent.Future
import java.util.function.BiConsumer

val GIT_JOB_CATEGORY = JobCategory.of("git").withName("Git")

interface GitService : SCMService {

    /**
     * Gets the Git configurator for a project
     */
    fun getGitConfiguratorAndConfiguration(project: Project): Pair<GitConfigurator, GitConfiguration>?

    /**
     * Tests if a project is correctly configured for Git.
     */
    fun isProjectConfiguredForGit(project: Project): Boolean

    /**
     * Tests if a branch is correctly configured for Git.
     */
    fun isBranchConfiguredForGit(branch: Branch): Boolean

    /**
     * Gets the configuration for a project.
     */
    fun getProjectConfiguration(project: Project): GitConfiguration?

    /**
     * Gets (if any) the pull request associated to a branch.
     *
     * Note that [GitService.getBranchAsPullRequest] must be used instead if you happen
     * to already have some information about the branch.
     *
     * @param branch Branch
     * @return Pull request or null is none
     */
    fun getBranchAsPullRequest(branch: Branch): GitPullRequest?

    /**
     * Is this branch a potential pull request?
     *
     * This method won't check if the branch associated pull request actually exists.
     *
     * @param branch Branch
     * @return Pull request status
     */
    fun isBranchAPullRequest(branch: Branch): Boolean

    /**
     * Given a branch and its Git configuration, gets its PR information.
     *
     * @param branch Branch
     * @param gitBranchConfigurationProperty Gt branch configuration
     * @return Pull request or null is none
     */
    fun getBranchAsPullRequest(branch: Branch, gitBranchConfigurationProperty: GitBranchConfigurationProperty?): GitPullRequest?

    /**
     * Gets the configuration for a branch
     *
     * @param branch Branch to check
     * @return Configuration.
     */
    fun getBranchConfiguration(branch: Branch): GitBranchConfiguration?

    /**
     * Gets an Ontrack branch for a given Git branch inside an Ontrack project.
     *
     * @param project Project where to look for the branch
     * @param branchName Name of the Git branch
     * @return Ontrack branch or `null` if not found
     */
    fun findBranchWithGitBranch(project: Project, branchName: String): Branch?

    /**
     * Launches the build/tag synchronisation for a branch
     */
    fun launchBuildSync(branchId: ID, synchronous: Boolean): Future<*>?

    /**
     * Loops over each correctly configured project.
     */
    fun forEachConfiguredProject(consumer: BiConsumer<Project, GitConfiguration>)

    /**
     * Loops over each correctly configured branch. Branch template definitions are excluded.
     */
    fun forEachConfiguredBranch(consumer: BiConsumer<Branch, GitBranchConfiguration>)


    /**
     * Loops over each correctly configured branch in a project. Branch template definitions are excluded.
     */
    fun forEachConfiguredBranchInProject(project: Project, consumer: (Branch, GitBranchConfiguration) -> Unit)

    /**
     * Gets information about an issue in a Git-configured project
     *
     * @param projectId ID of the project
     * @param key Display key of the issue
     * @return Issue & commit information about the issue in the project if available
     */
    fun getIssueProjectInfo(projectId: ID, key: String): OntrackGitIssueInfo?

    /**
     * Looks up a commit in the given `configuration`.
     *
     * @param id Commit long or short ID
     * @return The content of a commit if it exists, `null` otherwise.
     */
    fun lookupCommit(configuration: GitConfiguration, id: String): GitCommit?

    /**
     * Gets information about a commit in a Git-configured project.
     */
    fun getCommitProjectInfo(projectId: ID, commit: String): OntrackGitCommitInfo

    /**
     * Gets the list of remote branches, as defined under `ref/heads`.
     */
    fun getRemoteBranches(gitConfiguration: GitConfiguration): List<String>

    /**
     * Synchronises the Git repository attached to this project.
     *
     * @param project Project
     * @param request Sync request
     * @return Result. The synchronisation will occur asynchronously, but the acknowledgment returns
     * if the project did contain a Git configuration or not.
     */
    fun projectSync(project: Project, request: GitSynchronisationRequest): Ack

    /**
     * Synchronises the Git repository attached to this configuration using an asynchronous job.
     *
     * @param gitConfiguration Configuration to sync
     * @param request          Sync request
     * @return Result. The synchronisation will occur asynchronously, but the acknowledgment returns if the
     * synchronisation was actually launched.
     */
    fun sync(gitConfiguration: GitConfiguration, request: GitSynchronisationRequest): Future<*>?

    /**
     * Synchronises the Git repository attached to this configuration in a synchronous way. Same than [sync]
     * but without using a background job.
     *
     * This method is mostly used for testing.
     *
     * @param config Configuration to sync
     * @param project Associated project
     * @param listener Logger
     */
    fun syncProjectRepository(config: GitConfiguration, project: Project, listener: (message: String) -> Unit = { println(it) })

    /**
     * Gets the Git synchronisation information.
     *
     * @param project Project configured for Git
     * @return Synchronisation information
     */
    fun getProjectGitSyncInfo(project: Project): GitSynchronisationInfo

    fun scheduleGitBuildSync(branch: Branch, property: GitBranchConfigurationProperty)

    fun unscheduleGitBuildSync(branch: Branch, property: GitBranchConfigurationProperty)

    /**
     * Checks the log history and returns `true` if the token can be found.
     *
     * @param gitConfiguration Git configuration
     * @param token               Expression to be searched for
     * @return Result of the search
     */
    fun isPatternFound(gitConfiguration: GitConfiguration, token: String): Boolean

    /**
     * Gets the [GitCommit] associated to a build
     *
     * @param build Build to get the information for
     * @return Commit information or `null` if not found
     */
    fun getCommitForBuild(build: Build): IndexableGitCommit?

    /**
     * Sets the [GitCommit] information for a build.
     *
     * @param build Build to set the information for
     * @param commit Commit information
     */
    fun setCommitForBuild(build: Build, commit: IndexableGitCommit)

    /**
     * Collects and stores the [IndexableGitCommit]s for all builds of a branch.
     */
    fun collectIndexableGitCommitForBranch(branch: Branch, overrides: Boolean = true)

    /**
     * Collects and stores the [IndexableGitCommit]s for all builds of a branch.
     *
     * This is the optimized version for jobs running in the background.
     */
    fun collectIndexableGitCommitForBranch(branch: Branch,
                                           client: GitRepositoryClient,
                                           config: GitBranchConfiguration,
                                           overrides: Boolean,
                                           listener: JobRunListener)

    /**
     * Collects and stores the [IndexableGitCommit]s one build.
     */
    fun collectIndexableGitCommitForBuild(build: Build)

    /**
     * Converts a raw [GitCommit] into an annotated [GitUICommit]
     */
    fun toUICommit(gitConfiguration: GitConfiguration, commit: GitCommit): GitUICommit

    /**
     * Loops over the commits of a configuration
     */
    fun forEachCommit(gitConfiguration: GitConfiguration, code: (GitCommit) -> Unit)

    /**
     * Checks if the repository is ready to be used.
     */
    fun isRepositorySynched(gitConfiguration: GitConfiguration): Boolean

    /**
     * Gets the list of available issue export formats for this project
     */
    @Deprecated("Export formats are no longer issue service specific - will be removed in V5")
    fun getIssueExportFormats(project: Project): List<ExportFormat>

}
